tests/python/pants_test/engine/test_engine.py

```diff
@@ -237,6 +238,28 @@ def test_non_existing_root_fails_differently(self):
       str(cm.exception)
     )
 
+  def test_no_cache_exceptions(self):
+    error_message = 'this exception should not be cached!!'
+    def function_to_call() -> B:
+      raise ValueError(error_message)
+
+    @rule
+    def f(a: A) -> B:
+      return function_to_call()
+
+    scheduler = self.scheduler([f, RootRule(A)], include_trace_on_error=False)
+
+    a = A()
+
+    with self.assertRaises(ExecutionError):
+      scheduler.product_request(B, subjects=[a])
+
+    def function_to_call() -> B: # noqa: F811
```

Comment:
what does this `noqa: F811` comment mean?

Reply:
This allows lint to pass! Redefining any function is a lint error (it's also a mypy error, but we don't have that turned on here).